include/oneapi/tbb/detail/_task_handle.h

```diff
@@ -136,32 +136,56 @@ class task_dynamic_state {
 };
 #endif // __TBB_PREVIEW_TASK_GROUP_EXTENSIONS
 
+template <typename DerivedType>
+void destroy_function_task(task_handle_task* p, d1::small_object_allocator& alloc, const d1::execution_data* ed) {
+    if (ed) {
+        alloc.delete_object(static_cast<DerivedType*>(p), *ed);
+    } else {
+        alloc.delete_object(static_cast<DerivedType*>(p));
+    }
+}
+
 class task_handle_task : public d1::task {
-    std::uint64_t m_version_and_traits{};
+    // Pointer to the instantiation of destroy_function_task with the concrete derived type,
+    // used for correct destruction and deallocation of the task
+    using destroy_func_type = void (*)(task_handle_task*, d1::small_object_allocator&, const d1::execution_data*);
+
+    // Reuses the first std::uint64_t field (previously m_version_and_traits) to maintain backward compatibility
+    // The type of the first field remains std::uint64_t to preserve alignment and offset of subsequent member variables.
+    static_assert(sizeof(destroy_func_type) <= sizeof(std::uint64_t), "Cannot fit destroy pointer into std::uint64_t");
+    std::uint64_t m_destroy_func;
+
     d1::wait_tree_vertex_interface* m_wait_tree_vertex;
     d1::task_group_context& m_ctx;
     d1::small_object_allocator m_allocator;
 #if __TBB_PREVIEW_TASK_GROUP_EXTENSIONS
     std::atomic<task_dynamic_state*> m_dynamic_state;
 #endif
 public:
-    void finalize(const d1::execution_data* ed = nullptr) {
-        if (ed) {
-            m_allocator.delete_object(this, *ed);
+
+    void destroy(const d1::execution_data* ed = nullptr) {
```

Comment:
`destroy_this`? To make it clear that it's not `execution_data` being destroyed.